File: windows/Apps.cs
namespace Insight;

/// <summary>
/// What gets reported for a foreground process, and what gets thrown away.
///
/// Everything here is pure and covered by <see cref="SelfTest"/>, for the same
/// reason the extension's decision logic is: a bug in this file either records
/// something private or blocks the wrong thing.
/// </summary>
internal static class Apps
{
    /// <summary>
    /// Browsers are skipped entirely, because the extension already counts
    /// them.
    ///
    /// Counting both would double every minute a student spends on the web,
    /// and worse than double it: the extension would file that minute as
    /// distracted while this app filed the same minute as focused, so the
    /// distraction ratio — the number the whole insight engine turns on —
    /// would drift towards nonsense. A student with no extension loses their
    /// browser time here instead, which is a gap rather than a wrong answer.
    /// </summary>
    private static readonly HashSet<string> Browsers = new(StringComparer.OrdinalIgnoreCase)
    {
        "chrome", "msedge", "msedgewebview2", "firefox", "brave", "opera",
        "opera_gx", "vivaldi", "arc", "iexplore", "chromium", "thorium",
        "librewolf", "waterfox", "zen", "floorp",
    };

    /// <summary>
    /// Desktop apps reported under their website's name instead of their own.
    ///
    /// Focus Mode and the distraction figures are both defined by the
    /// student's blocklist, which is a list of hostnames — and a hostname is
    /// all the settings page can accept, because <c>normalizeSite</c> rejects
    /// anything that isn't domain-shaped. So the Spotify desktop app reported
    /// as "Spotify" would be untouchable: unblockable, and counted as focused
    /// time no matter what the student chose.
    ///
    /// Reporting it as <c>spotify.com</c> keeps one promise the rest of the
    /// app makes — that everything is measured against the student's own
    /// choices — and has the side effect of merging desktop and web time for
    /// the same service, which is what a student would expect anyway.
    ///
    /// Only apps whose hostname already appears in a block category are here.
    /// Anything else reports its own name.
    /// </summary>
    private static readonly Dictionary<string, string> Aliases = new(StringComparer.OrdinalIgnoreCase)
    {
        ["spotify"] = "spotify.com",
        ["discord"] = "discord.com",
        ["discordptb"] = "discord.com",
        ["discordcanary"] = "discord.com",
        ["steam"] = "steampowered.com",
        ["steamwebhelper"] = "steampowered.com",
        ["robloxplayerbeta"] = "roblox.com",
        ["epicgameslauncher"] = "epicgames.com",
        ["minecraftlauncher"] = "minecraft.net",
        ["minecraft"] = "minecraft.net",
        ["telegram"] = "telegram.org",
        ["whatsapp"] = "web.whatsapp.com",
        ["netflix"] = "netflix.com",
        ["twitch"] = "twitch.tv",
        ["tidal"] = "tidal.com",
        ["itunes"] = "music.apple.com",
        ["applemusic"] = "music.apple.com",
        ["primevideo"] = "primevideo.com",
        ["disney+"] = "disneyplus.com",
        ["hulu"] = "hulu.com",
        ["crunchyroll"] = "crunchyroll.com",
        ["vsco"] = "vsco.co",
        ["pinterest"] = "pinterest.com",
    };

    /// The longest a reported name may be — the server's own limit for the
    /// field. Truncating here means a strange binary produces a short label
    /// rather than a rejected batch that takes the whole minute down with it.
    private const int MaxNameLength = 253;

    /// <summary>
    /// What to report for a foreground process, or null to record nothing.
    /// </summary>
    /// <param name="processName">Executable name, without the .exe.</param>
    /// <param name="fileDescription">
    /// The binary's <c>FileDescription</c> resource — "Discord", "Notepad".
    /// Static metadata compiled into the executable, so unlike a window title
    /// it says what the app is and never what is open in it.
    /// </param>
    internal static string? Report(string? processName, string? fileDescription)
    {
        string process = (processName ?? "").Trim();
        if (process.Length == 0) return null;

        if (process.EndsWith(".exe", StringComparison.OrdinalIgnoreCase))
        {
            process = process[..^4];
        }
        if (process.Length == 0) return null;

        if (Browsers.Contains(process)) return null;

        if (Aliases.TryGetValue(process, out string? domain)) return domain;

        string name = Clean(fileDescription);
        if (name.Length == 0) name = Clean(process);
        if (name.Length == 0) return null;

        // A description can be an alias too — the Spotify Store build's
        // executable is not called "spotify".
        if (Aliases.TryGetValue(name, out string? byDescription)) return byDescription;

        return name.Length > MaxNameLength ? name[..MaxNameLength] : name;
    }

    /// <summary>
    /// Strip anything that would leak a path or break the display.
    ///
    /// A file description is free text set by whoever built the binary, so it
    /// is treated as untrusted: control characters out, path separators out,
    /// whitespace collapsed.
    /// </summary>
    private static string Clean(string? value)
    {
        if (string.IsNullOrWhiteSpace(value)) return "";

        var buffer = new System.Text.StringBuilder(value.Length);
        bool lastWasSpace = false;

        foreach (char c in value)
        {
            // Whitespace first, because a tab is a control character too and
            // dropping it outright would weld two words into one.
            if (char.IsWhiteSpace(c))
            {
                if (buffer.Length > 0 && !lastWasSpace) buffer.Append(' ');
                lastWasSpace = true;
                continue;
            }

            if (char.IsControl(c)) continue;
            if (c == '\\' || c == '/') continue;

            buffer.Append(c);
            lastWasSpace = false;
        }

        return buffer.ToString().Trim();
    }

    /// <summary>
    /// Does a reported name fall under a blocklist entry?
    ///
    /// The same rule as <c>matchesBlocklist</c> in <c>src/lib/blocklist.ts</c>
    /// and <c>isBlocked</c> in the extension. All three have to agree, because
    /// "blocked" and "counted as a distraction" are meant to be the same
    /// sentence — a student blocked from something that never appears in their
    /// distraction figures would have no idea what happened.
    /// </summary>
    internal static bool IsBlocked(string? name, IReadOnlyList<string> blocklist)
    {
        if (string.IsNullOrEmpty(name)) return false;

        string clean = name.ToLowerInvariant();
        if (clean.StartsWith("www.", StringComparison.Ordinal)) clean = clean[4..];

        foreach (string entry in blocklist)
        {
            if (clean == entry || clean.EndsWith("." + entry, StringComparison.Ordinal))
            {
                return true;
            }
        }
        return false;
    }
}

File: mac/Sources/Insight/Apps.swift
import Foundation

/// What gets reported for the app in front, and what gets thrown away.
///
/// The Windows twin of this file is `windows/Apps.cs` and the two must agree,
/// because a student with both machines should see one set of figures rather
/// than two dialects of the same day. Everything here is pure and covered by
/// `SelfTest`.
enum Apps {

    /// Browsers are skipped entirely, because the extension already counts
    /// them.
    ///
    /// Counting both would double every minute on the web, and worse, the
    /// extension would file that minute as distracted while this filed it as
    /// focused, so the distraction ratio the insight engine turns on would
    /// drift towards nonsense. A student with no extension loses their browser
    /// time here, which is a gap rather than a wrong answer.
    ///
    /// Matched on bundle identifier, which is exact. The Windows version has
    /// to match on executable names and guess a little.
    static let browsers: Set<String> = [
        "com.apple.Safari",
        "com.apple.SafariTechnologyPreview",
        "com.google.Chrome",
        "com.google.Chrome.canary",
        "com.microsoft.edgemac",
        "org.mozilla.firefox",
        "org.mozilla.firefoxdeveloperedition",
        "com.brave.Browser",
        "com.operasoftware.Opera",
        "com.vivaldi.Vivaldi",
        "company.thebrowser.Browser",
        "com.thebrowser.dia",
        "org.chromium.Chromium",
        "net.waterfox.waterfox",
        "app.zen-browser.zen",
    ]

    /// Apps reported under their website's name instead of their own.
    ///
    /// Focus Mode and the distraction split are both defined by the student's
    /// blocklist, which is a list of hostnames, `normalizeSite` won't accept
    /// anything else. So the Spotify app reported as "Spotify" would be
    /// untouchable: unblockable, and counted as focused time whatever the
    /// student chose. Reporting it as `spotify.com` keeps the whole app on one
    /// principle, that everything is measured against their own choices.
    ///
    /// Only apps whose hostname already appears in a block category are here.
    static let aliases: [String: String] = [
        "com.spotify.client": "spotify.com",
        "com.hnc.Discord": "discord.com",
        "com.hnc.DiscordPTB": "discord.com",
        "com.hnc.DiscordCanary": "discord.com",
        "com.valvesoftware.steam": "steampowered.com",
        "com.roblox.RobloxPlayer": "roblox.com",
        "com.epicgames.launcher": "epicgames.com",
        "com.mojang.minecraftlauncher": "minecraft.net",
        "ru.keepcoder.Telegram": "telegram.org",
        "org.telegram.desktop": "telegram.org",
        "net.whatsapp.WhatsApp": "web.whatsapp.com",
        "com.netflix.Netflix": "netflix.com",
        "tv.twitch.desktop": "twitch.tv",
        "com.tidal.desktop": "tidal.com",
        "com.apple.Music": "music.apple.com",
        "co.vsco.vscox": "vsco.co",
        "com.pinterest.mac": "pinterest.com",
    ]

    /// Names that map even when the bundle identifier is unfamiliar, a
    /// sideloaded or re-signed build of the same app.
    static let aliasesByName: [String: String] = [
        "spotify": "spotify.com",
        "discord": "discord.com",
        "steam": "steampowered.com",
        "roblox": "roblox.com",
        "netflix": "netflix.com",
        "twitch": "twitch.tv",
        "telegram": "telegram.org",
        "whatsapp": "web.whatsapp.com",
        "minecraft": "minecraft.net",
    ]

    /// The server's own limit for the field. Truncating here means a strange
    /// app produces a short label rather than a rejected batch that takes the
    /// whole minute down with it.
    static let maxNameLength = 253

    /// What to report for the app in front, or nil to record nothing.
    ///
    /// - Parameters:
    ///   - bundleId: e.g. `com.apple.TextEdit`. Exact, and free, unlike a
    ///     window title, reading it needs no Accessibility permission, which
    ///     is the permission this app deliberately never asks for.
    ///   - name: the app's localised name, e.g. "TextEdit".
    static func report(bundleId: String?, name: String?) -> String? {
        if let bundleId, browsers.contains(bundleId) { return nil }

        // Our own windows must not be counted as an app they chose to use.
        // Both sides have to exist for this to mean anything: run outside a
        // bundle, `swift run`, or the self-test, our own identifier is nil,
        // and a bare `==` would then swallow every app that hasn't got one.
        if let bundleId, let own = Bundle.main.bundleIdentifier, bundleId == own {
            return nil
        }

        if let bundleId, let domain = aliases[bundleId] { return domain }

        let cleaned = clean(name)
        if cleaned.isEmpty { return nil }

        if let domain = aliasesByName[cleaned.lowercased()] { return domain }

        return String(cleaned.prefix(maxNameLength))
    }

    /// Strip anything that would leak a path or break the display.
    ///
    /// An app's name comes from its bundle and is set by whoever built it, so
    /// it is treated as untrusted: control characters out, path separators
    /// out, whitespace collapsed. Note that whitespace is checked first, a
    /// tab is a control character too, and dropping it outright welds two
    /// words into one. The Windows version shipped that bug for an hour.
    static func clean(_ value: String?) -> String {
        guard let value else { return "" }

        var out = ""
        var lastWasSpace = false

        for character in value {
            if character.isWhitespace {
                if !out.isEmpty && !lastWasSpace { out.append(" ") }
                lastWasSpace = true
                continue
            }

            if character.unicodeScalars.allSatisfy({ CharacterSet.controlCharacters.contains($0) }) {
                continue
            }
            if character == "/" || character == "\\" { continue }

            out.append(character)
            lastWasSpace = false
        }

        return out.trimmingCharacters(in: .whitespaces)
    }

    /// Does a reported name fall under a blocklist entry?
    ///
    /// The same rule as `matchesBlocklist` in `src/lib/blocklist.ts`, `isBlocked`
    /// in the extension, and `Apps.IsBlocked` in the Windows app. All four have
    /// to agree, because "blocked" and "counted as a distraction" are meant to
    /// be the same sentence.
    static func isBlocked(_ name: String?, blocklist: [String]) -> Bool {
        guard let name, !name.isEmpty else { return false }

        var clean = name.lowercased()
        if clean.hasPrefix("www.") { clean = String(clean.dropFirst(4)) }

        return blocklist.contains { clean == $0 || clean.hasSuffix("." + $0) }
    }
}
